refactor(devtools): improve directive explorer filtering

hawkgs opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

- [x] Feature


## What is the current behavior?

The current directive explorer filtering UI (highlighting of the matched nodes) is a bit obtrusive and does not indicate which part of the filter query matches with the node unlike Chrome DevTools.

## What is the new behavior?

https://github.com/user-attachments/assets/d78b35e2-01b9-4bf7-8e36-b722340defed

- Highlight only the matched part of the node name/text
- Select the first match by default
- Show the matches count along with the currently selected one
- Minor UI fixes, performance optimizations and code cleanup. 
- Some tests for the new `tree-node` (formerly part of `directive-forest`) and `filter-input`. The old functionality in `directive-forest` hasn't been tested.

Note: The Regex filtering has been removed in favor of simple string matching. The existing UI doesn't have filter options like enabling/disabling case sensitivity or Regex opt out ability, so it doesn't make sense to have Regex filtering enabled by default since it might cause problems in some situations.

cc @dgp1130 

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

hawkgs committed:
refactor(devtools): improve directive explorer filtering

- Highlight only the matched part of the text
- Select the first match by default
- Show the matches count along with the currently selected one
- Minor UI fixes and performance optimizations

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-forest/filter/filter.component.html:
```diff
@@ -12,7 +12,14 @@
+      @if (currentMatch(); as currentMatch) {
+        {{ currentMatch }} of {{ matchesCount() }}
+      } @else {
+        {{ matchesCount() }} match{{ matchesCount() !== 1 ? 'es' : '' }}
```
Consider: `{{ matchesCount() === 1 ? 'match' : 'matches' }}`.

Deduplicating "match" is probably more confusing than helpful.

Could also consider ``{{ matchesCount() === 1 ? "1 match" : `${matchesCount()} matches` }}`` which is probably closer to what this would be for an i18n app.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-forest/filter/filter.component.spec.ts:
```diff
@@ -0,0 +1,68 @@
+
+  it('should render if there is a single text match', () => {
+    let count = getMatchesCountEl();
+    expect(count).toBeFalsy();
```
Nit: If `getMatchesCountEl()` is obsoleted anytime the component re-renders, it might be simpler to just not store it in a variable where not needed.

```typescript
expect(getMatchesCountEl()).toBeFalsy();

fixture.componentRef.setInput('matchesCount', 1);
fixture.detectChanges();

expect(getMatchesCountEl()!.nativeElement.innerText).toBe('1 match');
```

That way you don't have to worry about updating `count` or accidentally asserting on an old value.

hawkgs replied:
:+1: Ah, didn't see it. Introduced the function to avoid repetition after the variables had already been in place.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-forest/directive-forest.component.html:
```diff
@@ -1,89 +1,21 @@
+  (prevMatched)="navigateMatchedNode('prev')"
+  [matchesCount]="matchesCount()"
+  [currentMatch]="currentlyMatchedIndex() + 1"
+/>
```
Consider: This might be a more complicated rearchitecture than it's worth, but I'm wondering if there's a way to have `<ng-filter />` take a little more responsibility when it comes to managing the matched text?

Any consumer of `<ng-filter />` would need to reimplement `matchesCount` and `currentMatch` and duplicate a sizable chunk of logic. In particular, now that `<ng-filter />` has more responsibility in showing the number of matches, it maybe deserves to own a little more of the logic.

I'm not 100% how this would work, but a couple ideas to consider:
1. `(filter)="..."` could output a filter _function_ rather than a string of text. `directive-forest` could call that function to filter the data.
2. `directive-forest` could pass in the `dataSource` to `<ng-filter />` which would output the highlighting and selected match information which feeds into `<ng-tree-node />` below.
3. Have `directive-forest` inject `<ng-filter />` as a `viewChild` and imperatively collaborate with it.
4. Make a separate service/controller class owned by `directive-forest` and passed in to `<ng-filter />` to have the two components collaborate more imperatively without requiring _direct_ knowledge of each other.

Ultimately I'm ok with what you have here, just pointing out some alternatives to think through and consider if they might make the collaboration between these components a bit simpler and make `<ng-filter />` more self-contained as an abstraction.

hawkgs replied:
```typescript
// filter.component.ts 

export type FilterFn = (source: string) => {
  start: number;
  end: number;
} | null;

emitFilter(event: Event): void {
  const filterStr = (event.target as HTMLInputElement).value;

  const filterFn: FilterFn = (target: string) => {
    if (!filterStr) {
      return null;
    }
    const startIdx = target.toLowerCase().indexOf(filterStr.toLowerCase());

    if (startIdx > -1) {
      return {
        start: startIdx,
        end: startIdx + filterStr.length,
      };
    }
    return null;
  };

  this.filter.emit(filterFn);
}
```

Something like this? This is point one. This will retain the reusability characteristics of the filter, if we decide to ever use it somewhere else in the app. The difference is not that big but I can see how this can become more useful when we enhance the filtering from plain string search to something more sophisticated.

2. Passing the data source to the filter seems a bit odd to me in terms of how the data flows through the component tree. It also effectively turns `filter` to `directive-forest-filter`. If we want to retain the reusability of `filter`, we have to introduce some form of an abstraction layer that'll come with additional overhead and complexity.
3. Not sure how this will work. At least I am unable to imagine the benefits. Maybe I need more details.
4. I guess `directive-forest` and `filter` are bound to have direct knowledge of each other since `directive-forest` is the user? From what I interpret, this should be similar to (2.) but the data source is kept in that service. 

Note: The results highlighting functionality is bound to directive explorer, so I don't see how this can effectively change unless we go crazy with the abstractions.

dgp1130 replied:
I'll defer to you what the best option is here (even if just keeping it the same to what you have now). I mainly just wanted to point out some additional options and make sure we're thinking through the abstraction and collaboration model we're creating with `directive-forest` and `ng-filter`. I agree there's a lot of nuanced trade-offs here.

hawkgs replied:
Okay, thanks! I feel that your first proposal checks the most boxes:

1. Delegates the core filtering logic to the filter
2. Retains reusability
3. Doesn't require additional data abstractions
4. Provides a good platform for further extensions that will be abstracted from the user (`directive-forest`) – whether it's case sensitivity, Regex support or parsing of the input value for a more sophisticated filtering.

I'll let you know when I introduce the change to the PR :+1:

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-forest/tree-node/tree-node.component.html:
```diff
@@ -0,0 +1,66 @@
+) {
+  <div class="hydration-error">
+    @if (hydration.expectedNodeDetails) {
+      <div>Expected Dom:</div>
```
Nit: Typically I see "DOM" all uppercase.

hawkgs replied:
I'll limit the change to `tree-node`. I guess we'll need a larger scale refactoring for the rest of the components.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-forest/tree-node/tree-node.component.spec.ts:
```diff
@@ -0,0 +1,147 @@
+
+    const name = fixture.debugElement.query(By.css('.node-name'));
+
+    expect(name.nativeElement.innerText).toEqual('app-test[TooltipDirective]');
```
Consider: Test with at least two directives names to confirm that rendering.

(Admittedly I think it's just a string, so maybe doesn't matter too much here.)

hawkgs replied:
I'll add few more tests.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-forest/tree-node/tree-node.component.ts:
```diff
@@ -0,0 +1,165 @@
+          this.elementRef.nativeElement,
+          'padding-left',
+          (this.node().level + 1) * PADDING_LEFT_STEP + 'px',
+        ),
```
Consider: Do we need `Renderer2` for this? Can we bind `[style.padding-left]` to a `computed` with this value?

hawkgs replied:
Yes, we need it since we are applying the style to the host and we can't do it via the template.

hawkgs replied:
Didn't see the comment but I was just updating that after your last comment about the `@Host*` decorators. Still not used to using `host` 😄

dgp1130 reviewed (approved):
Thanks for looking into this, definitely appreciate the in-depth tests!

Feel free to ignore any comments which apply to the previous implementation, it's a bit tricky to see what code is actually new and what's just moved around.

I had some thoughts on the abstraction model of `<ng-filter />` and the UX of searching for multiple directives, but their both minor concerns and I think what you have already is pretty reasonable. Up to you if there's anything worth changing there.

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-forest/directive-forest-utils.ts:
```diff
@@ -35,3 +35,12 @@ export const parentCollapsed = (
+  if (cmp && cmp.isElement) {
+    return `<${node.name}/>`;
+  }
+  return node.name + (node.directives ? `[${node.directives}]` : '');
```
Frankly, I haven't paid much attention to this detail but I am following the [existing design](https://github.com/angular/angular/blob/main/devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-forest/directive-forest.component.html#L43). I thought that directives are space separated (i.e. `component[dir1 dir2 dir3]` but it turns out they are comma separated.

So, you suggest putting each directive in its own brackets? In terms of readability I think that `a[x y z]` is better compared to `a[x][y][z]` but semantically it seems wrong, I guess.

dgp1130 replied:
I thought it was already `[x][y][z]` TBH. I don't feel strongly about it, happy to defer to whatever format you think is most readable.

hawkgs replied:
It wasn't but I did the change so it's `a[x][y][z]` now. :+1:

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-forest/directive-forest.component.ts:
```diff
@@ -309,109 +244,145 @@ export class DirectiveForestComponent {
-        indexesOfMatchedNodes.push(i);
+      const node = this.dataSource.data[i];
+      const fullName = getFullNodeName(node).toLowerCase();
+      const startIdx = fullName.indexOf(filterText.toLowerCase());
```
Yes, it will result in no matches. Currently, it's a plain string search as you noted.

How does (1.) differ from the current functionality? Or you mean that `foo baz`, for instance, should match `foo[bar][baz]`? If that's the case, it seems that (2.) is the way to go. Without any markers like the brackets it becomes hard to infer what's what and that will interfere with the string search. I think that in its core, the filter should still retain the plain string search as baseline functionality since we don't know how the directive explorer names will evolve in the future.

dgp1130 replied:
I agree 1. isn't that different from a basic text search. We could restrict that `[]` don't appear in the filter, but I don't think that would provide much meaningful value, so I'm inclined to let users enter `a[b]` even if it depends on arbitrary directive ordering.

I don't feel too strongly and don't expect the case of searching for `a[b]` when `a[c][b]` is in the tree to come up too often. Up to you how to handle it.

hawkgs replied:
I like the idea! I can implement it. Given that the current change doesn't degrade the filter, would you be okay to do the implementation in a separate PR since this one might become a bit bulky and I guess it will be easier for you to review? I see this feature as an incrementation of the existing change.

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-forest/directive-forest.component.ts:
```diff
@@ -64,41 +68,38 @@ export class DirectiveForestComponent {
+  readonly highlightIdInTreeFromElement = signal<number | null>(null);
+  readonly matchedNodes = signal<Map<number, NodeTextMatch>>(new Map()); // Node index, NodeTextMatch
+  readonly matchesCount = computed(() => this.matchedNodes().size);
+  readonly currentlyMatchedIndex = signal<number>(-1);
```
Should we follow this convention for the methods as well? It becomes a bit wordy.

hawkgs committed:
fixup! refactor(devtools): improve directive explorer filtering

- Address most of the PR comments
- Add more tests
- Drop HostBindings and HostListeners
- Change how the filter component interacts with its parent

JeanMeche commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-forest/tree-node/tree-node.component.ts:
```diff
@@ -0,0 +1,165 @@
+    afterRenderEffect({write: () => this.handleMatchedText()});
+  }
+
+  @HostBinding('class.selected')
```
The recommendation today is to use the host property on the decorator. 

v20 will land typesafety and highlighting for it.

mmalerba commented:
This PR was merged into the repository by commit a1deb87a8e5a994537af28e9ae7dbde3ac4ba087.

The changes were merged into the following branches: main
